018-custom-id-support.md

```diff
@@ -0,0 +1,128 @@
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Developers want to name their resources with their own custom ID that they can easily query with. This will also help make query URL understandable and pretty.
+
+**What is the context or background in which this problem exists?**
+
+At the moment the IDs for all the resources are auto generated by Appwrite. Developers are forced to use the generated IDs in their applications to make queries. The generated ids are random and not easily understandable.
+
+**Once the proposal is implemented, how will the system change?**
+
+Every resource creation API will have an optional **id** attribute. Whenever **id** is not provided Appwrite will still generate the IDs as it is already doing. IDs will not be modifiable once the resource is created.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail keeping in mind the following:
+
+- Its interaction with other parts of the system is clear
+- It is reasonably clear how the contribution would be implemented
+- Dependencies on libraries, tools, projects or work that isn't yet complete
+- New API routes that need to be created or modifications to the existing routes (if needed)
+- Any breaking changes and ways in which we can ensure backward compatibility.
+- Use Cases
+- Goals
+- Deliverables
+- Changes to documentation
+- Ways to scale the solution
+
+Ensure that you include examples, code-snippets etc. to allow the community to understand the proposed solution. **It would be best if the examples use naming conventions that you intend to use during the actual implementation so that changes can be suggested early on during the development.**
+
+Write your answer below.
+
+-->
+Resources that can have Custom IDs
+- Projects
+- Collections
+- Documents
+- Buckets
+- Files
+- Users
+- Function tags
+- Tasks
+- Webhooks
+
+The create resource endpoints for above resources will have a first required **id** parameter where users can send their custom id following the guidelines for validation. In every endpont, if the **id** parameter is received empty (required for backward compatibility with existing SDKs) or pre defined string, Appwrite will still generate the ids as it is already doing.
+
+We can use the existing key validator to validate the custom id as well.
+
+### Some guidelines for custom ids
+
+- Can be at most 32 characters
+- Can be Alphanumeric
+- Can contain `_` but not at the begining of the id
+- Cannot contain any other special characters
+
+### Initial implementation
+
+The first implementation will not contain following features:
+- TTL
+- Maximum Number of files
```

Comment:
Is this a leftover from the Storage Bucket RFC? 😄

Reply:
Ah thanks. removed